Origin: upstream, https://github.com/moby/moby/pull/35739/commits/fb170206ba12752214630b269a40ac7be6115ed4
From fb170206ba12752214630b269a40ac7be6115ed4 Mon Sep 17 00:00:00 2001
From: Derek McGowan <derek@mcgstyle.net>
Date: Mon, 2 Apr 2018 10:48:34 -0700
Subject: [PATCH] Set format in archiver

Prevent changing the tar output by setting the format to
PAX and keeping the times truncated.
Without this change the archiver will produce different tar
archives with different hashes with go 1.10.
The addition of the access and changetime timestamps would
also cause diff comparisons to fail.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
---
 pkg/archive/archive.go      | 9 +++++++++
 pkg/containerfs/archiver.go | 5 +++++
 2 files changed, 14 insertions(+)

--- a/engine/pkg/archive/archive.go
+++ b/engine/pkg/archive/archive.go
@@ -16,8 +16,9 @@
 	"runtime"
 	"strconv"
 	"strings"
 	"syscall"
+	"time"
 
 	"github.com/docker/docker/pkg/fileutils"
 	"github.com/docker/docker/pkg/idtools"
 	"github.com/docker/docker/pkg/ioutils"
@@ -359,8 +360,12 @@
 	hdr, err := tar.FileInfoHeader(fi, link)
 	if err != nil {
 		return nil, err
 	}
+	hdr.Format = tar.FormatPAX
+	hdr.ModTime = hdr.ModTime.Truncate(time.Second)
+	hdr.AccessTime = time.Time{}
+	hdr.ChangeTime = time.Time{}
 	hdr.Mode = fillGo18FileTypeBits(int64(chmodTarEntry(os.FileMode(hdr.Mode))), fi)
 	name, err = canonicalTarName(name, fi.IsDir())
 	if err != nil {
 		return nil, fmt.Errorf("tar: cannot canonicalize path: %v", err)
@@ -1157,8 +1162,12 @@
 			hdr, err := tar.FileInfoHeader(srcSt, "")
 			if err != nil {
 				return err
 			}
+			hdr.Format = tar.FormatPAX
+			hdr.ModTime = hdr.ModTime.Truncate(time.Second)
+			hdr.AccessTime = time.Time{}
+			hdr.ChangeTime = time.Time{}
 			hdr.Name = filepath.Base(dst)
 			hdr.Mode = int64(chmodTarEntry(os.FileMode(hdr.Mode)))
 
 			if err := remapIDs(archiver.IDMappingsVar, hdr); err != nil {
--- a/engine/pkg/containerfs/archiver.go
+++ b/engine/pkg/containerfs/archiver.go
@@ -5,8 +5,9 @@
 	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"time"
 
 	"github.com/docker/docker/pkg/archive"
 	"github.com/docker/docker/pkg/idtools"
 	"github.com/docker/docker/pkg/system"
@@ -137,8 +138,12 @@
 			hdr, err := tar.FileInfoHeader(srcSt, "")
 			if err != nil {
 				return err
 			}
+			hdr.Format = tar.FormatPAX
+			hdr.ModTime = hdr.ModTime.Truncate(time.Second)
+			hdr.AccessTime = time.Time{}
+			hdr.ChangeTime = time.Time{}
 			hdr.Name = dstDriver.Base(dst)
 			if dstDriver.OS() == "windows" {
 				hdr.Mode = int64(chmodTarEntry(os.FileMode(hdr.Mode)))
 			} else {
